packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,462 @@
+import { QueryPoolsResponse } from '@kyvejs/types/lcd/kyve/query/v1beta1/pools';
+import { delay, getLogger, IBlock } from '@subql/node-core';
+import axios, { AxiosResponse } from 'axios';
+import { remove } from 'lodash';
+import { BlockContent } from '../../indexer/types';
+import { formatBlockUtil, LazyBlockContent } from '../cosmos';
+import { isTmpDir } from '../project';
+import { BundleDetails } from './kyveTypes';
+
+const BUNDLE_TIMEOUT = 10000; //ms
+const POLL_TIMER = 3; // sec
+const MAX_COMPRESSION_BYTE_SIZE = 2 * 10 ** 9;
+
+const parseDecimal = (value: string) => parseInt(value, 10);
+
+const logger = getLogger('kyve');
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { version: packageVersion } = require('../../../package.json');
+
+interface KyveBundleData {
+  value: { block: any; block_results: any };
+  key: string;
+}
+
+export class KyveApi {
+  private cachedBundleDetails: BundleDetails[] = [];
+
+  private constructor(
+    private readonly storageUrl: string,
+    private readonly tmpCacheDir: string,
+    private readonly poolId: string,
+    private readonly lcdClient: KyveLCDClientType,
+  ) {}
+
+  static async create(
+    chainId: string, // chainId for indexing chain
+    endpoint: string,
+    storageUrl: string,
+    kyveChainId: SupportedChains,
+    tmpCacheDir: string,
+  ): Promise<KyveApi> {
+    if (!isTmpDir(tmpCacheDir)) {
+      throw new Error('File cache directory must be a tmp directory');
+    }
+
+    const lcdClient = new KyveSDK(kyveChainId, {
+      rpc: endpoint,
+    }).createLCDClient();
+
+    const poolId = await KyveApi.fetchPoolId(chainId, lcdClient);
+
+    return new KyveApi(storageUrl, tmpCacheDir, poolId, lcdClient);
+  }
+
+  private static async fetchPoolId(
+    chainId: string,
+    lcdClient: KyveLCDClientType,
+  ): Promise<string> {
+    const poolsResponse =
+      (await lcdClient.kyve.query.v1beta1.pools()) as unknown as QueryPoolsResponse;
+
+    for (const p of poolsResponse.pools) {
+      try {
+        const config = JSON.parse(p.data.config);
+        if (config.network === chainId) {
+          return p.id; // Return the matching pool ID
+        }
+      } catch (error) {
+        throw new Error(
+          `Error parsing JSON for pool with id ${p.id}: ${error}`,
+        );
+      }
+    }
+
+    throw new Error(`${chainId} is not available on Kyve network`);
+  }
+
+  private decodeBlock(block: JsonRpcSuccessResponse): BlockResponse {
+    return Responses.decodeBlock({
+      id: 1,
+      jsonrpc: '2.0',
+      result: block,
+    });
+  }
+
+  private decodeBlockResult(
+    blockResult: JsonRpcSuccessResponse,
+  ): BlockResultsResponse {
+    return Responses.decodeBlockResults({
+      id: 1,
+      jsonrpc: '2.0',
+      result: blockResult,
+    });
+  }
+
+  private getBundleFilePath(id: string): string {
+    return path.join(this.tmpCacheDir, `bundle_${id}.json`);
```

Comment:
Is it worth including the chain id in the file name? What happens if 2 projects end up having the same directory?